Refactor photos in redux

jmalm opened the pull request:
Fixes #314.

Photos are now stored in `store.photos.photosFlat` and `store.photos.photosGroupedByDate` (in addition to `photosFlat`, if the fetched set of photos comes grouped by date, otherwise empty), instead of having a separate place for each different kind of photo set, in different parts of the store (store.photos, store.albums, store.search). This makes it much easier to do things like update the favorite icons in photo list views (which this PR also takes care of).

I didn't look at places albums in this PR, simply because I currently have no test data. All other photo sets _should_ be updated and working correctly.

This PR builds on https://github.com/LibrePhotos/librephotos-frontend/pull/23, which is a small PR that should be merged first.

Commit:
Fix broken thumbnail in ModalPhotosShare view

jmalm committed:
Merge branch 'LibrePhotos:dev' into dev

Commit:
Refactor how photos are stored in the Redux state

Fixes #314, but more importantly makes it easier to update the current
photo list, because the same photo list is used for all Photos views.
Also makes favorite icons in photo list views update without need
for refreshing the page.

derneuere commented on src/reducers/photosReducer.js:
```diff
@@ -212,120 +259,142 @@ export default function reducer(
+         *
+         * Alternatives:
+         * 1. Let the unfavorited photo remain in the view (but turn off the star icon).
+         * 2. Remove the unfavorited photo from the view and close the lightbox.
```
I think I prefer the second alternative too.
To close the lightbox, you have to change the state in PhotoListView of "lightboxShow"

https://github.com/LibrePhotos/librephotos-frontend/blob/263e69f063d1b45314dfaba0887a8f86ceabf0f3/src/components/photolist/PhotoListView.js#L41

jmalm replied:
Right. Do you have a suggestion of how to accomplish that? I wasn't able to figure out a good way...

jmalm replied:
Code updated to close lightbox when unfavoriting in Favorited view.

derneuere commented on src/layouts/SearchView.js:
```diff
@@ -1,38 +1,41 @@
-        )
-    }
+  render() {
+    // I don't know when the alternative titles would be shown, since the
```
I think this was the title when you would navigate to /search without a query parameter, and I probably regressed on that while I was refactoring PhotoListView.

I think it would make more sense, if search without a query parameter would redirect you.

jmalm replied:
Right. I agree. A redirect to '/', I assume?

jmalm replied:
Code updated to redirect to / when the user navigates to /search without a query.

derneuere commented on src/components/sharing/ModalPhotosShare.js:
```diff
@@ -102,7 +102,6 @@ export class ModalPhotosShare extends Component {
                 <SecuredImageJWT
                   key={"selected_image" + image}
                   height={40}
```
I think I would prefer it if the picture is rectangular like here:
https://github.com/LibrePhotos/librephotos-frontend/blob/263e69f063d1b45314dfaba0887a8f86ceabf0f3/src/components/album/ModalAlbumEdit.js#L100-L109

jmalm replied:
This was actually already merged to dev in https://github.com/LibrePhotos/librephotos-frontend/pull/23

Looks like the SecuredImageJWT is making this problematic. I removed the width setting because the thumbnails of landscape photos were squeezed horizontally (although square_thumbnails urls are used). I'm not sure what the SecuredImageJWT is for...

Maybe create a separate issue for making these thumbnails square again?

derneuere replied:
I created two issues https://github.com/LibrePhotos/librephotos/issues/332 and https://github.com/LibrePhotos/librephotos/issues/331 :+1:

derneuere commented:
I checked the location views and couldn't find any error. 

The only thing that is currently not working for me is the favorite overlay. It just does not want to appear. Does it work on your system? It could be that my dependencies are somehow not working 😅

jmalm commented:
Oh? Do you mean that the favorite icon never shows up in photo list views? Like this:

![image](https://user-images.githubusercontent.com/3975811/131231003-763eef2a-6b11-4319-a834-5c9a2f341884.png)

That is pretty essential in this PR, so let's smoke this out... I don't think that any dependencies were added for this. Have the favorite icons never showed up for you in photo list views?

derneuere commented:
Not sure what happened. I rebuild the docker image and now it works 😅

Commit:
Redirect to / when user navigates to /search

Commit:
Close lightbox when unfavoriting

When in the Favorites view and unmarking a photo
as favorite, close the lightbox if it is shown.

derneuere commented:
Now it all looks good! :+1:
